File: Programmers/전력망둘로나누기_0428.kt
import kotlin.math.min
import kotlin.math.abs
class Solution {
    var answer: Int = Int.MAX_VALUE
    var tree = Array(0) {BooleanArray(0)}
    var checked = BooleanArray(0) {false}
    
    fun dfs(cur: Int, n: Int): Int {
        var child = 1
        checked[cur] = true
        for (next in 1..n) {
            if(tree[cur][next] && checked[next] == false){
                child += dfs(next, n)
            }
        }
        answer = min(answer, abs(n - child - child))
        return child
    }

    fun solution(n: Int, wires: Array<IntArray>): Int {
        tree = Array(n+1) {BooleanArray(n+1)}
        checked = BooleanArray(n+1) {false}
        for(wire in wires) {

            tree[wire[0]][wire[1]] = true
            tree[wire[1]][wire[0]] = true
        }

        dfs(2, n)

        return answer
    }
}

File: Programmers/전력망둘로나누기_0428.py
def solution(n, wires):
    answer = n
    tree = [[] for _ in range(n + 1)]

    for a, b in wires:
        tree[a].append(b)
        tree[b].append(a)

    visited = [False] * (n + 1)
    child = [0] * (n + 1)

   # 노드 마다 자식 개수 구하기
    def child_count(node):
        nonlocal visited, answer
        visited[node] = True
        child_cnt = 1 # 자기자신 포함
        
        for next in tree[node]:
            if not visited[next]:
                visited[next] = True
                child_cnt += child_count(next)
        answer = min(answer,abs(n-child_cnt-child_cnt))
        return child_cnt #(현재 노드의 자식개수 + 자기자신) 부모에게 돌려주기

    child_count(1)

    return answer